src/lib/Syntax.hs

```diff
@@ -792,54 +882,97 @@ instance BindsUVars a => BindsUVars (Maybe a) where
   boundUVars Nothing = mempty
   boundUVars (Just x) = boundUVars x
 
-nameAsEnv :: Name -> UVars
-nameAsEnv v = v@>()
-
 -- === Expr free variables and substitutions ===
 
+-- In the safer-names system we use Typeable to make this an open sum type, but
+-- here all the cases are explicit.
+data AnyBinderInfo =
+   AtomBinderInfo Type BinderInfo
+ | DataDefName  DataDef
+ | TyConName    DataDefName
+ | DataConName  DataDefName Int
+ | SuperclassName ClassDefName Int Atom
+ | MethodName     ClassDefName Int Atom
+ | LocalUExprBound
+ | ImpBound
+ | TrulyUnknownBinder
+   deriving (Show, Generic)
+
+-- Just documentation for now, but it'll be a distinct type
+-- with safer-names
+type DataDefName = Name
+type ClassDefName = DataDefName
+
 data BinderInfo =
         LamBound (ArrowP ())
         -- TODO: make the expression optional, for when it's effectful?
         -- (or we could put the effect tag on the let annotation)
       | PatBound
       | LetBound LetAnn Expr
-      | DataBoundDataCon DataDef Int
-      | DataBoundTypeCon DataDef
       | PiBound
       | UnknownBinder
         deriving (Show, Generic)
 
-type SubstEnv = Env Atom
-type Bindings = Env (Type, BinderInfo)
+data SubstVal e = SubstVal e
+                | Rename Name
+
+type SubstEnv = Env (SubstVal Atom)
+type Bindings = Env AnyBinderInfo
 type Scope = Bindings  -- when we only care about the names, not the payloads
 type ScopedSubstEnv = (SubstEnv, Bindings)
 
-scopelessSubst :: Subst a => SubstEnv -> a -> a
+-- TODO: we could add a lot more structure for querying by dict type, caching, etc.
+data SynthCandidates = SynthCandidates
+  { lambdaDicts       :: [Atom]
+  , superclassGetters :: [Atom]
+  , instanceDicts     :: [Atom] }
```

Comment:
> missing some very important scoping information

Then it's a good opportunity to use name space types! Here's what it will look like in safer-names:

```haskell
data EvaluatedModule n =
  EvaluatedModule
    :: BindingsFrag n l
    -> SynthCandidates l
    -> SourceMap l
    -> EvaluatedModule n

data SynthCandidates n = SynthCandidates
  { lambdaDicts       :: [Atom n]
  , superclassGetters :: [Atom n]
  , instanceDicts     :: [Atom n] }
```

> Do the atoms use module's internal namespace? 

Yes (see types above)

> How do they relate to the source map? 

They're complementary. Modules export source names, but they also export un-named class instances. These SynthCandidates are those instances.

> Is their type the only important information which is why there are no named binders for them?

There are no names because they're anonymous instances.

> Why do we separate the three categories?

We treat them differently during dictionary synthesis (as we always have). We could have a single `[(Atom, SynthCandidateFlavor)]` list instead, but this is a bit more efficient. Note that the `lambdaDicts` list should be empty, because there are no top-level lambda binders. The field is there because we use the same data structure in the dict synthesis monad.